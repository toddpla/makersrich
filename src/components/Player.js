import React, { Component } from 'react';
import { SPRITE_SIZE } from '../constants'
import PlayerImg from "../assets/player.png"
import { connect } from 'react-redux'
import { collectItem, digTile, unDigTile } from '../actions/map'
import { startAddInventoryItem, startUpdatePlayer, startCreditPlayer } from '../actions/auth'
import { startSendNewsfeedMessage } from '../actions/newsfeed'
import SpinningCoin from '../assets/spinning_coin_16px.gif'
import MessageForm from './messages/MessageForm'
import MessageFeed from './messages/MessageFeed'


export class Player extends Component {

  constructor(props) {
    super(props)
    this.state = {
      inInventory: false,
      messageOnFocus: false
      level: this.getPlayerLevel(props)
    }
  }

  getPlayerLevel = (props) => {
    var questions = 0

    if (props.player.questions !== undefined ) {
      questions = Object.keys(props.player.questions).length
    }

    var sessionQuestions = this.props.player.sessionQuestions.length
    var level =  (questions + sessionQuestions) / 5
    return Math.floor(level)
  }

  updatePlayerLevel = (props) => {
    this.setState({
      level: this.getPlayerLevel(props)
    })
  }

  handleKeyDown = (e) => {
    if (!this.state.inInventory && !this.props.notOnMap) {
      switch(e.keyCode) {
        // left key
        case 37:
          return this.props.handleMovement({ left: this.props.player.left - SPRITE_SIZE, top: this.props.player.top })
        // up key
        case 38:
          return this.props.handleMovement({ top: this.props.player.top - SPRITE_SIZE, left: this.props.player.left  })
        // right key
        case 39:
          return this.props.handleMovement({ left: this.props.player.left + SPRITE_SIZE, top: this.props.player.top  })
        // down key
        case 40:
          this.updatePlayerLevel(this.props)
          return this.props.handleMovement({ top: this.props.player.top + SPRITE_SIZE, left: this.props.player.left  })
        case 69:
          return this.attemptDig(this.props.player.left, this.props.player.top)
        case 73:
          this.updatePlayerLevel(this.props)
          return this.props.handlePopupInventory()
        case 82:
          return this.props.handlePopupRPS()
        case 81:
          return this.props.checkSign(this.props.player.left, this.props.player.top)
        case 88:
          return this.props.handlePopupInstructions()
        default:
          console.log(e.keyCode);
      }
    } else {
        return this.popUpHandleKeyDown(e)
    }
  }

  popUpHandleKeyDown(e) {
    if (this.state.inInventory ) {
      switch(e.keyCode) {
        case 73:
          this.setState({ inInventory: false })
          return this.props.closeModal()
        default:
          console.log(e.keyCode);
      }
    }
  }

  coinAnimation = () => {
    var timeout = 2;
    this.setState({coinCollected: true})
    setTimeout(this.hideAnimation, 1600)
  }

  hideAnimation = () => {
    this.setState({coinCollected: false})
  }

  possibleCash(chance = Math.random()) {
    var coinCount = 0
    if (chance < 0.01) {
      coinCount += 25
    } else if (chance < 0.1) {
      coinCount += 5
    } else if (chance < 0.25) {
      coinCount += 1
    }
    if (coinCount !== 0) {
      this.props.startCreditPlayer(coinCount)
      this.coinAnimation()
    }
  }

  digDatDing = (x, y) => {
    this.possibleCash()

    var dug = document.createElement("div")
    dug.setAttribute('class', 'dug-up-tile')
    dug.setAttribute('id', x+y)
    dug.setAttribute('style', `left:${x}px; top:${y}px` )
    document.getElementsByClassName("map")[0].appendChild(dug)
    setTimeout(this.unDigDatDing, 5000, x, y)
  }

  unDigDatDing = (x, y) => {
    const tile = this.props.map.minable.filter((object) => object.x === x && object.y === y)[0]
    const element = document.getElementById(x+y)
    element.parentNode.removeChild(element)
    this.props.unDigTile(tile)
  }

  attemptDig = (x, y) => {
    const tile = this.props.map.minable.filter((object) => object.x === x && object.y === y)[0]
    if (tile.visible === false) {
      console.log('you can dig');
      this.props.digTile(tile)
      this.digDatDing(x, y)
    } else {
      console.log('no digging!');
    }

    const item = this.props.map.collectables.filter((object) => object.x === x && object.y === y)[0]
    this.checkItem(item)
  }

  checkItem(item) {
    if (item !== undefined) {
      this.props.collectItem(item)
      this.props.startAddInventoryItem(item.type, item)
      this.setState({inPopUp: true})
      this.props.startSendNewsfeedMessage(`${this.props.player.displayName.split(' ')[0]} found a ${item.type}!`)
      this.props.handlePopupMessage(`You found a ${item.type}!`)
    }
  }

  componentDidUpdate() {
    window.addEventListener('keydown', (e) => {
      if (!this.state.messageOnFocus) {
        e.preventDefault()
        this.handleKeyDown(e)
      }
    })
  }

  handleOnFocus = () => {
    this.setState({messageOnFocus: true})
  }

  render() {
      if (this.state.coinCollected) {
        var button = <div id="coinContainer"
                  style={{
                    position: 'absolute',
                    width: '16px',
                    height: '16px',
                    backgroundImage: `url(${SpinningCoin})`,
                    backgroundPosition: 'center',
                    zIndex: 1,
                    top: this.props.player.top - 16,
                    left: this.props.player.left
                  }}
                 >
        </div>
      } else { var button = undefined }

      return (
      <div>
      {button}

      <div id="player"
        style={{
          position: 'absolute',
          width: '16px',
          top: this.props.player.top,
          left: this.props.player.left,
          height: '16px',
          backgroundPosition: 'center',
          backgroundImage: `url(${PlayerImg})`
        }}
      >
      </div>
      <MessageForm handleOnFocus={this.handleOnFocus}  style= {{zIndex: 0, width: '40px', height: '40px'}}/>
      <MessageFeed  />
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  map: state.map,
  player: state.auth
})

const mapDispatchToProps = (dispatch) => ({
  collectItem: (item) => dispatch(collectItem(item)),
  digTile: (tile) => dispatch(digTile(tile)),
  unDigTile: (tile) => dispatch(unDigTile(tile)),
  startUpdatePlayer: (updates) => dispatch(startUpdatePlayer(updates)),
  startAddInventoryItem: (itemRef, item) => dispatch(startAddInventoryItem(itemRef, item)),
  startSendNewsfeedMessage: (message) => dispatch(startSendNewsfeedMessage(message)),
  startCreditPlayer: (amount) => dispatch(startCreditPlayer(amount))
})

export default connect(mapStateToProps, mapDispatchToProps)(Player);
