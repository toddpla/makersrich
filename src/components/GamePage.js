import React, { Component } from 'react';
import Player from './Player'
import { MapProvider, Map } from 'react-tiled'
import { connect } from 'react-redux'
import { updatePlayer } from '../actions/players'
import { MAX_HEIGHT, MAX_WIDTH, SPRITE_SIZE } from '../constants'
import styled from "styled-components";
import Modal from 'react-modal'
import Quiz from './quiz/Quiz'
import Inventory from './Inventory/Inventory'

import mapJson from '../POWLevel1.json'

const customStyles = {
  content : {
    top                   : '50%',
    left                  : '50%',
    right                 : 'auto',
    bottom                : 'auto',
    marginRight           : '-50%',
    transform             : 'translate(-50%, -50%)',
    backgroundColor       : 'rgba(255, 0, 0, 0)'
  }
};

const AppWrapper = styled.div`
  height: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
  background-color: #1c1117;
`

Modal.setAppElement('#root')

class GamePage extends Component {

  constructor() {
    super();

    this.state = {
      modalIsOpen: false,
      modalComponenet: 'undefined'
    };

    this.openModal = this.openModal.bind(this);
    this.closeModal = this.closeModal.bind(this);
  }

  openModal(popUpMessage) {
    this.setState({
      modalIsOpen: true,
      ...popUpMessage
    });
  }

  closeModal() {
    this.setState({modalIsOpen: false});
  }

  handleMovement = (player, updates) => {
    console.log(player, updates)
    if (!this.checkBoundaries(updates) && this.checkImpassable(updates)) {
      this.props.updatePlayer(player, updates)
      this.forceUpdate()
    }
  }

  checkImpassable = (updates) => {
    const x = updates.left
    const y = updates.top

    const impassablePos = mapJson.layers[1].objects.filter((object) => object.x === x && object.y === y)[0]
    return (impassablePos !== undefined) ? false : true
  }

  checkBoundaries = (updates) => {
    if (updates === undefined) { return }
    return (updates.left < 0 || updates.top < 0 || updates.left > MAX_WIDTH - SPRITE_SIZE
            || updates.top > MAX_HEIGHT - SPRITE_SIZE )
  }

  handlePopupQuiz = () => {
    this.openModal({modalComponenet: <Quiz />})
  }
  handlePopupInventory = () => {
    this.openModal({modalComponenet: <Inventory />})
  }

  render() {
    return (
      <div>
      <MapProvider style={{margin: "auto"}}  mapUrl={process.env.PUBLIC_URL + "/assets/POWLevel1.json"}>
       <AppWrapper>
        <Map style={{ transform: "scale(1)", position: 'relative' }}>
          <div>
            {this.props.players.map((player, i) => <Player key={i} player={player} handleMovement={this.handleMovement} /> )}
          </div>
        </Map>
       </AppWrapper>
      </MapProvider>
      <button onClick={this.openModal}>Open Modal</button>
      <button onClick={this.handlePopupQuiz}>Open Quiz</button>
      <button onClick={this.handlePopupInventory}>Open Invetory</button>
        <Modal
          isOpen={this.state.modalIsOpen}
          onAfterOpen={this.afterOpenModal}
          onRequestClose={this.closeModal}
          style={customStyles}
          contentLabel={this.state.modalTitle}
        >
          {this.state.modalComponenet}
          <button onClick={this.closeModal}>close</button>
        </Modal>
        </div>
    );
  }
}

const mapStateToProps = (state) => ({
  players: state.players
})

const mapDispatchToProps = (dispatch) => ({
  updatePlayer: (player, direction) => dispatch(updatePlayer(player, direction))
})

export default connect(mapStateToProps, mapDispatchToProps)(GamePage);
