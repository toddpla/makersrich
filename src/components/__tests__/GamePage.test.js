import React from 'react'
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme'
import { GamePage, AppWrapper } from '../GamePage'
import { Player } from '../Player'

let startUpdatePlayer = jest.fn()
let impassablePos = jest.fn()
let handlePopupMessage = jest.fn()
let openModal = jest.fn()
let setState = jest.fn()

let props = {
  player: 'player',
  opponents: [],
  map: {
    impassable: [{x: 15, y: 30}],
    portals: [{x: 15, y: 30, name: 'Shop'}],
    signs: [{x: 15, y: 14, properties: [{value: 'This is a sign'}] }]
  }
}

let updates = {left: 0, top: 0}

test('gamepage renders correctly', () => {
  const gamePage = shallow(<GamePage {...props}/>)
  expect(gamePage).toMatchSnapshot();
});


test('AppWrapper renders correctly', () => {
  const appWrapper = renderer.create(<AppWrapper/>).toJSON();
  expect(appWrapper).toMatchSnapshot();
});

test('#handleMovement calls action to update player state', () => {
  let wrapper = shallow( <GamePage {...props} startUpdatePlayer={startUpdatePlayer}/> )
  const instance = wrapper.instance()
  instance.handleMovement(updates)
  expect(startUpdatePlayer).toHaveBeenLastCalledWith(updates)
})

// test('#handleMovement calls action to update player state', () => {
//   let wrapper = shallow( <GamePage {...props} startUpdatePlayer={startUpdatePlayer}/> )
//   const instance = wrapper.instance()
//   instance.handleMovement(updates)
//   expect(startUpdatePlayer).toHaveBeenLastCalledWith(updates)
// })

test('#checkBoundaries returns undefined with no updates', () => {
  let wrapper = shallow( <GamePage {...props}/> )
  const instance = wrapper.instance()
  expect(instance.checkBoundaries(undefined)).toEqual()
})

test('#checkPortal returns false with no portal', () => {
  let wrapper = shallow( <GamePage {...props}/> )
  const instance = wrapper.instance()
  expect(instance.checkPortal(0,0)).toEqual(false)
})

test('#checkPortal returns true with portal', () => {
  let wrapper = shallow( <GamePage {...props}/> )
  const instance = wrapper.instance()
  expect(instance.checkPortal(15,30)).toEqual('Shop')
})

test('#checkSign returns false when no sign', () => {
  let wrapper = shallow( <GamePage {...props}/> )
  const instance = wrapper.instance()
  expect(instance.checkSign(0, 0)).toEqual(false)
})

// test('#checkSign call a function when there is a sign', () => {
//   let wrapper = shallow( <GamePage {...props}/> )
//   const instance = wrapper.instance()
//   instance.checkSign(15, 30)
//   expect(handlePopupMessage).toHaveBeenLastCalledWith('This is a sign')
// })

test('#checkImpassable returns true when not impassable', () => {
  let wrapper = shallow( <GamePage {...props}/> )
  const instance = wrapper.instance()
  expect(instance.checkImpassable({left: 0, top: 0})).toEqual(true)
})

test('#checkImpassable returns false when impassable', () => {
  let wrapper = shallow( <GamePage {...props}/> )
  const instance = wrapper.instance()
  expect(instance.checkImpassable({left: 15, top: 30})).toEqual(false)
})
