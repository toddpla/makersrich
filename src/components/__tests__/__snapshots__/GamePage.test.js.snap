// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gamepage renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <GamePage
    egg={
      Object {
        "left": 0,
        "top": 0,
      }
    }
    map={
      Object {
        "impassable": Array [
          Object {
            "x": 15,
            "y": 30,
          },
        ],
        "portals": Array [
          Object {
            "name": "Shop",
            "x": 15,
            "y": 30,
          },
        ],
        "signs": Array [
          Object {
            "properties": Array [
              Object {
                "value": "This is a sign",
              },
            ],
            "x": 15,
            "y": 14,
          },
        ],
      }
    }
    opponents={Array []}
    player="player"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Map />,
        <Connect(Player)
          checkSign={[Function]}
          closeModal={[Function]}
          handleMovement={[Function]}
          handlePopupInstructions={[Function]}
          handlePopupInventory={[Function]}
          handlePopupMessage={[Function]}
          notOnMap={false}
          onFocus={false}
          player="player"
        />,
        Array [],
        <Connect(Egg) />,
        <div>
          <ControlPanel
            handleOffFocus={[Function]}
            handleOnFocus={[Function]}
            handlePopupLeaderboard={[Function]}
          />
        </div>,
        <Modal
          ariaHideApp={false}
          bodyOpenClassName="ReactModal__Body--open"
          closeTimeoutMS={0}
          isOpen={false}
          onRequestClose={[Function]}
          parentSelector={[Function]}
          portalClassName="ReactModalPortal"
          role="dialog"
          shouldCloseOnEsc={true}
          shouldCloseOnOverlayClick={true}
          shouldFocusAfterRender={true}
          shouldReturnFocusAfterClose={true}
          style={
            Object {
              "content": Object {
                "background": "none",
                "border": "none",
                "bottom": "auto",
                "left": "50%",
                "marginRight": "-50%",
                "right": "auto",
                "top": "50%",
                "transform": "translate(-50%, -50%)",
              },
            }
          }
        >
          undefined
          <div
            className="modal-button"
            onClick={[Function]}
          />
        </Modal>,
      ],
      "id": "game-wrapper",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "checkSign": [Function],
          "closeModal": [Function],
          "handleMovement": [Function],
          "handlePopupInstructions": [Function],
          "handlePopupInventory": [Function],
          "handlePopupMessage": [Function],
          "handlePopupRPS": undefined,
          "notOnMap": false,
          "onFocus": false,
          "player": "player",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <ControlPanel
            handleOffFocus={[Function]}
            handleOnFocus={[Function]}
            handlePopupLeaderboard={[Function]}
          />,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "handleOffFocus": [Function],
            "handleOnFocus": [Function],
            "handlePopupLeaderboard": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "ariaHideApp": false,
          "bodyOpenClassName": "ReactModal__Body--open",
          "children": Array [
            "undefined",
            <div
              className="modal-button"
              onClick={[Function]}
            />,
          ],
          "closeTimeoutMS": 0,
          "contentLabel": undefined,
          "isOpen": false,
          "onAfterOpen": undefined,
          "onRequestClose": [Function],
          "parentSelector": [Function],
          "portalClassName": "ReactModalPortal",
          "role": "dialog",
          "shouldCloseOnEsc": true,
          "shouldCloseOnOverlayClick": true,
          "shouldFocusAfterRender": true,
          "shouldReturnFocusAfterClose": true,
          "style": Object {
            "content": Object {
              "background": "none",
              "border": "none",
              "bottom": "auto",
              "left": "50%",
              "marginRight": "-50%",
              "right": "auto",
              "top": "50%",
              "transform": "translate(-50%, -50%)",
            },
          },
        },
        "ref": null,
        "rendered": Array [
          "undefined",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "modal-button",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": "div",
          },
        ],
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Map />,
          <Connect(Player)
            checkSign={[Function]}
            closeModal={[Function]}
            handleMovement={[Function]}
            handlePopupInstructions={[Function]}
            handlePopupInventory={[Function]}
            handlePopupMessage={[Function]}
            notOnMap={false}
            onFocus={false}
            player="player"
          />,
          Array [],
          <Connect(Egg) />,
          <div>
            <ControlPanel
              handleOffFocus={[Function]}
              handleOnFocus={[Function]}
              handlePopupLeaderboard={[Function]}
            />
          </div>,
          <Modal
            ariaHideApp={false}
            bodyOpenClassName="ReactModal__Body--open"
            closeTimeoutMS={0}
            isOpen={false}
            onRequestClose={[Function]}
            parentSelector={[Function]}
            portalClassName="ReactModalPortal"
            role="dialog"
            shouldCloseOnEsc={true}
            shouldCloseOnOverlayClick={true}
            shouldFocusAfterRender={true}
            shouldReturnFocusAfterClose={true}
            style={
              Object {
                "content": Object {
                  "background": "none",
                  "border": "none",
                  "bottom": "auto",
                  "left": "50%",
                  "marginRight": "-50%",
                  "right": "auto",
                  "top": "50%",
                  "transform": "translate(-50%, -50%)",
                },
              }
            }
          >
            undefined
            <div
              className="modal-button"
              onClick={[Function]}
            />
          </Modal>,
        ],
        "id": "game-wrapper",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "checkSign": [Function],
            "closeModal": [Function],
            "handleMovement": [Function],
            "handlePopupInstructions": [Function],
            "handlePopupInventory": [Function],
            "handlePopupMessage": [Function],
            "handlePopupRPS": undefined,
            "notOnMap": false,
            "onFocus": false,
            "player": "player",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <ControlPanel
              handleOffFocus={[Function]}
              handleOnFocus={[Function]}
              handlePopupLeaderboard={[Function]}
            />,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "handleOffFocus": [Function],
              "handleOnFocus": [Function],
              "handlePopupLeaderboard": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "ariaHideApp": false,
            "bodyOpenClassName": "ReactModal__Body--open",
            "children": Array [
              "undefined",
              <div
                className="modal-button"
                onClick={[Function]}
              />,
            ],
            "closeTimeoutMS": 0,
            "contentLabel": undefined,
            "isOpen": false,
            "onAfterOpen": undefined,
            "onRequestClose": [Function],
            "parentSelector": [Function],
            "portalClassName": "ReactModalPortal",
            "role": "dialog",
            "shouldCloseOnEsc": true,
            "shouldCloseOnOverlayClick": true,
            "shouldFocusAfterRender": true,
            "shouldReturnFocusAfterClose": true,
            "style": Object {
              "content": Object {
                "background": "none",
                "border": "none",
                "bottom": "auto",
                "left": "50%",
                "marginRight": "-50%",
                "right": "auto",
                "top": "50%",
                "transform": "translate(-50%, -50%)",
              },
            },
          },
          "ref": null,
          "rendered": Array [
            "undefined",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "modal-button",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": "div",
            },
          ],
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
